rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+optimize its decoder to choose the most efficient Arrow type.
+
+#### Conversion of string features
+
+Arrow’s string arrays (`BinaryArray`) have a different memory layout than
+TensorFlow’s string Tensors, even with
+[`tensorflow::tstring`](https://github.com/tensorflow/community/blob/master/rfcs/20190411-string-unification.md).
+There is always some overhead in conversion, but with `tensorflow::tstring` a
+Tensor of `string_view`s is possible, thus the overhead will be a function of
+the number of strings being converted, instead of the lengths of the strings.
+
+#### TF APIs for conversions
+
+In TF 1.x we will use np.ndarray as a bridge as Arrow has zero-copy conversion
+to numpy’s ndarrays. (not for string arrays).
+
+Starting from TF 2.x, we will be able to create EagerTensors from Python
+memoryview(s) so that strings can be covered.
+
+## TFMD Schema
+
+[The TFMD Schema](https://github.com/tensorflow/metadata/blob/master/tensorflow_metadata/proto/v0/schema.proto)
+is a pipeline-level artifact and in the scope of this proposal, it may serve two
+purposes:
+
+*   To provide optional inputs to the parsing logic for optimizations.
+*   To carry user’s intents of converting data to TF feedables.
+
+The two purposes don’t have to be served in the following cases:
+
+*   TFDV should not require a schema to work and it does not need TF feedables.
+*   Some TFXIO implementation may not need the schema for either purposes.
+
+Therefore the TFMD schema is optional, and a TFXIO implementation:
+
+*   should guarantee that the `BeamSource()`can return a valid
+    `PCollection[RecordBatch]` without a schema.
+    *   Other interfaces may raise an error when a schema was not provided.
+*   does not have to require a TFMD schema for all its interfaces to work.
+
+## (TensorFlow) Trainer integration
+
+For TFX to freely choose the storage format for training examples for a user, we
+**cannot** expose file-based or record-based interface to that user in the TF
+trainer, because:
+
+*   the user might not know how to open those files.
+*   there might not be an efficient representation of a “record” (this is true
+    for columnar storage formats like Apache Parquet) but only an efficient
+    representation of a batch of records.
+
+Thus we propose that to most users, the TF Trainer only exposes a handle to a
+`tf.data.Dataset` of parsed (composite) Tensors.
+
+Each `TFXIO` implementation will implement a `TFDataset()` interface to return
+such a `tf.data.Dataset`. This dataset contains logically a set of batched
+(composite) Tensors that are of the same type as the corresponding
+`TensorAdapter()` would return for a `RecordBatch`. See
+[this section](#recommended-way-of-implementing-a-tfxio) about how to minimize
+the code needs to be written for a new `TFXIO` implementation.
+
+The `TFDataset()` interface will accept common knobs that a user may need to
+tweak:
+
+*   Batch size
+*   Random shuffle
```

Comment:
There are many, many knobs. Does it make sense to just pass all kwargs along?

Reply:
There will likely to be many switches to control the visible behavior of the Dataset, but ideally no knobs (that tunes performance).

We can look at tf.data.experimental.make_batched_features_dataset() [1] as it is quite similar to the proposed API here (i.e. it is a one-stop shop for a Dataset of tensors, in contrast of chaining tf.data transforms)

Many kwargs passed into it are knobs that we shouldn't be exposing to our users:
- prefetch_buffer_size
- reader_num_threads
- parser_num_threads

TFX is a more controlled environment (the pipeline is aware of what type of machine it is running on), and I would argue that TFX knows better how to set those knobs than the user does.

The rest of kwargs are controlling the behavior: batch_size, num_epochs, shuffle, shuffle_buffer_size, shuffle_seed, sloppy_ordering and yes.. they will have to be passed along, because most of them are expressing record-level constraints / properties of the dataset. The most efficient way of implementing them depends on the storage format. I'm not sure we can afford converting a columnar representation of a batch of records into individual records (that's unbatch() if implemented naively).

[1] https://github.com/tensorflow/tensorflow/blob/abb92231a1a0bffcd68ce09f051851ce761cc7ca/tensorflow/python/data/experimental/ops/readers.py#L750

Reply:
On the other hand, aligning with how the Generic Trainer approaches the I/O abstraction problem, which is to provide a dataset that contains "records" to allow record-based operations like shuffling, we could provide a similar interface that provide "record" based access, the only difference is that the records here are 1-row Arrow RecordBatches. Record-based operations are allowed, but at the same time, "records" are opaque to the user. In the end, they will need to use an op provided by us to rebatch those records into one big RecordBatch.